cpp/src/streaming/core/fanout.cpp

```diff
@@ -0,0 +1,430 @@
+            return idx != InvalidIdx;
+        });
+
+        auto it = std::ranges::begin(filtered_view);  // first valid idx
+        auto end = std::ranges::end(filtered_view);  // end idx
+
+        if (it == end) {
+            // no valid indices, so all send tasks are in an invalid state
+            return true;
+        }
+
+        auto [min_it, max_it] = std::minmax_element(it, end);
+        last_completed_idx = *min_it;
+        latest_processed_idx = *max_it;
+
+        return latest_processed_idx == state.recv_messages.size();
+    });
+
+    // if both last_completed_idx and latest_processed_idx are invalid, it means that all
+    // send tasks are in an invalid state.
+    co_return !(last_completed_idx == InvalidIdx && latest_processed_idx == InvalidIdx);
+}
+
+/**
+ * @brief Process input messages and notify send tasks to copy & send messages.
+ *
+ * @param ctx The context to use.
+ * @param ch_in The input channel to receive messages from.
+ * @param state The state of the unbounded fanout.
+ * @return A coroutine representing the task.
+ */
+Node unbounded_fo_process_input_task(
+    Context& ctx,
+    std::shared_ptr<Channel> ch_in,
+    std::shared_ptr<UnboundedFanoutState> state
+) {
+    ShutdownAtExit ch_in_shutdown{ch_in};
+    StateInputDoneAtExit state_closer{state};
+    co_await ctx.executor()->schedule();
+    auto& logger = ctx.logger();
+
+    logger.trace("Scheduled process input task");
+
+    // index of the first message to purge
+    size_t purge_idx = 0;
+
+    // input_done is only set by this task, so reading without lock is safe here
+    while (!state->input_done) {
+        size_t last_completed_idx = InvalidIdx, latest_processed_idx = InvalidIdx;
+
+        if (!co_await wait_for_data_request(
+                *state, last_completed_idx, latest_processed_idx
+            ))
+        {
+            break;  // waiting returned an invalid state, so we need to break
+        }
+
+        // receive a message from the input channel
+        auto msg = co_await ch_in->receive();
+
+        {  // relock mtx to update input_done/recv_messages
+            auto lock = co_await state->mtx.scoped_lock();
+            if (msg.empty()) {
+                state->input_done = true;
+            } else {
+                state->recv_messages.emplace_back(std::move(msg));
+            }
+        }
+
+        // notify send_tasks to copy & send messages
+        co_await state->data_ready.notify_all();
+
+        // purge completed send_tasks. This will reset the messages to empty, so that
+        // they release the memory, however the deque is not resized. This guarantees
+        // that the indices are not invalidated. intentionally not locking the mtx
+        // here, because we only need to know a lower-bound on the last completed idx
+        // (ch_next_idx values are monotonically increasing)
+        while (purge_idx + 1 < last_completed_idx) {
+            state->recv_messages[purge_idx].reset();
```

Comment:
Do we need to purge here, could the last `unbounded_fo_send_task` not move the message out of `recv_messages`?